src/util/convert_expr.h

```diff
@@ -9,20 +9,19 @@ Author: Diffblue ltd. All Rights Reserved
 /// \file
 /// Abstraction Refinement Loop
 
-#ifndef CPROVER_SOLVERS_REFINEMENT_EXPR_CAST_H
-#define CPROVER_SOLVERS_REFINEMENT_EXPR_CAST_H
+#ifndef CPROVER_UTIL_CONVERT_EXPR_H
+#define CPROVER_UTIL_CONVERT_EXPR_H
 
 #include <util/optional.h>
 
 /// Convert exprt to a specific type. Remove empty optional if
 /// conversion cannot be performed
 /// Generic case doesn't exist, specialize for different types accordingly
-/// TODO: this should go to util
 template<typename T>
-struct expr_cast_implt final { };
+struct from_expr_implt final { };
```

Comment:
Apparently this is more than just a plain file renaming? It's also a very confusing one, as there already are functions called `from_expr`.

Reply:
`from_expr` functions convert an expression to something else (which is also the case here). Here the type we are converting to is given as template argument so the meaning should be clear. But I'm open to suggestions for better names.

Reply:
To be honest, I'm not very happy with the "convert" terminology. You are doing (partly) checked type casts, nothing more and nothing less. To me, "convert" suggests that there is some sort of transformation going on (which would be true of the existing `from_expr` code).

Reply:
In that vein: `checked_cast_...` maybe? Also call the file `checked_expr_cast.h`?

Reply:
Checked typecast is more what is done in expr_cast.h and it is a bit different from these functions.
In particular `expr_checked_cast` is already used there.

Reply:
Honestly, I have no idea what this code is doing. The file used to be named "expr_cast.h", hence I made my inferences. There is a single comment in this file and having read that comment block three times now I still have no clue what it is trying to tell me (what is a "specific type", what is "empty" about an "optional" and how does an "optional" come into play when we are talking about `exprt`?). You are now saying it's "a bit different"? How much is "a bit"?

Reply:
(I wrote the original version). I wanted to follow the naming convention of `boost::lexical_cast`, only rather than throwing exceptions, I wanted to return empty optional to signal "bad conversion". It was developed independently of `expr_dynamic_cast`. But then the `expr_dynamic_cast` was merged (after `expr_cast` was already used locally in `string_refinement`). (`expr_dynamic_cast` and `expr_try_dynamic_cast` return references and pointers respectively).

Personally, I'd leave it as "expr_cast" and let it be the function which converts all sorts of `exprts` to various types in optional, and returning empty optional on error.

expr_cast.h doesn't actually have a function called "expr_cast", so it could be declared there.

Reply:
You are making it sound like there are multiple somewhat competing implementations? That's interesting.

Maybe it would help (this might be what you are suggesting in your last sentence) to have all of this in one file? Then at least there would be more context to help understand the code.

Reply:
There is a single implementation for casting between `exprt`s with two different interfaces - throw on error or return `nullptr` on error (in `util/expr_cast.h`). This function however casts from `exprt` to an integer, string, or whatever other types we can reliably cast to. But since `exprt` is cheap to copy, it could also use the same pattern - `return optionalt<derived_exprt>`. Why it wasn't used in the first place - I have no idea.

Reply:
Thank you very much for providing further detail -- what does it even mean to "cast" an `exprt` in this way? Anyway at least I believe I understand what the code is supposed to do. I have no idea how this code ever made it into the code base in its current form, but then that's a different question that the approvers of #1241 need to answer.

Anyway, it seems `util/arith_tools.h` seems to be the place for most of them.

Reply:
I don't think that conversion from `exprt` to string or from one `exprt` to another is an arithmetic operation. It made its way into the code, as it provides saner way of doing those things compared to 15+ line dance with tons of `INVARIANTS` every time someone wanted to cast const_exprt to its C++ equivalent.

Reply:
> I don't think that conversion from exprt to string or from one exprt to another is an arithmetic operation.

Indeed, I generalised too quickly - the integer conversion ones should go in `arith_tools.h`, `exprt` to `*_exprt` should go in `expr_cast.h`, and `exprt` to string I have no idea what that is.

> The only "problem" is the name, which is debatable.

The problems certainly include the complete absence of documentation as well as poor choice of placement in the file system. Is it saner? I don't know, because I don't know what it's supposed to do (as there is no documentation).

Reply:
Re: `expr_dynamic_cast` and returning pointers vs. `optionalt`, I believe @reuk was the person advocating pointers because the syntax to use them is briefer -- `optionalt<reference_wrapper<T>>` requires an explicit `.get()` operation a little too often.

Reply:
@smowton Before expr_dynamic_cast, there was this, which was intended to return `optional<T>` instead of `optional<reference_wrapper<T>>`. Old casts (`to_****_expr`) were doing that, because it's not inefficient, so I followed, as it was the safest and most concise way of doing those things.